rkyv/src/util/aligned_vec.rs

```diff
@@ -2,35 +2,53 @@
 use ::alloc::{alloc, boxed::Box, vec::Vec};
 use core::borrow::{Borrow, BorrowMut};
 use core::{
-    fmt,
+    cmp, fmt, hash,
     ops::{Deref, DerefMut, Index, IndexMut},
-    ptr::NonNull,
     slice,
+    slice::SliceIndex,
 };
 #[cfg(feature = "std")]
-use std::{alloc, io};
+use std::io;
+
+use crate::{AlignedBytes, Archive, Deserialize, Serialize};
+
+const ALIGNMENT: usize = 16;
+
+// TODO: replace when int_roundings feature is stabilized
+const fn div_ceiling(dividend: usize, divisor: usize) -> usize {
+    (dividend + divisor - 1) / divisor
+}
 
 /// A vector of bytes that aligns its memory to 16 bytes.
+///
+/// The alignment also applies to [`ArchivedAlignedVec`], which is useful for aligning opaque bytes inside of an archived data
+/// type.
+///
+/// ```
+/// # use rkyv::{AlignedVec, Archive, Serialize};
+/// #
+/// #[derive(Archive, Serialize)]
+/// struct HasAlignedBytes {
+///     pub bytes: AlignedVec,
+/// }
+///
+/// let mut bytes = AlignedVec::new();
+/// bytes.extend_from_slice(&[1, 2, 3]);
+/// let ser_foo = rkyv::to_bytes::<_, 0>(&HasAlignedBytes { bytes }).unwrap();
+/// let arch_foo = unsafe { rkyv::archived_root::<HasAlignedBytes>(&ser_foo) };
+/// assert_eq!(arch_foo.bytes.as_slice(), &[1, 2, 3]);
+/// assert_eq!(arch_foo.bytes.as_ptr().align_offset(16), 0);
+/// ```
+#[derive(Archive, Clone, Deserialize, Serialize)]
+#[archive(crate = "crate")]
 pub struct AlignedVec {
-    ptr: NonNull<u8>,
-    cap: usize,
+    bytes: Vec<AlignedBytes<ALIGNMENT>>,
```

Comment:
Because `Vec` contains its own `ptr, cap, len` triple, this will bump `size_of::<AlignedVec>()` up from 24 to 32. It's not necessarily a big problem but it is wasted space since we have redundant information in two `len`s.